src/metabase/api/card.clj

```diff
@@ -365,6 +365,11 @@ saved later when it is ready."
       (when timed-out?
         (schedule-metadata-saving result-metadata-chan <>)))))
 
+(defn- check-valid-card-info [{:keys [dataset dataset_query]}]
+  (when (and dataset
+             (not (every? #(contains? #{"card" "snippet"} %) (get-in dataset_query [:native :template-tags]))))
+    (throw (ex-info (tru "A model made from a native SQL question cannot have a variable or field filter.") {:status-code 400}))))
+
```

Comment:
Maybe there's a better place to put this?

Reply:
My pick is we make it to `metabase.models.card` with the `pre-insert` and `pre-update` hooks since we're currently doing a bunch of checks there https://github.com/metabase/metabase/blob/82748cda662a0a439f53dce73c0e139d57c1bfc8/src/metabase/models/card.clj#L210-L215